### ISSUE TITLE
[ENHANCEMENT] Support setting `exclude` list in the storage-network setting

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#3478]
**Is your enhancement related to a problem? Please describe.**

Sometimes the user wants to exclude some IPs from the storage network subnet.
But the storage-network setting on GUI doesn't have a field to enter excluded CIDRs.

<img width="1176" alt="storage-network" src="https://user-images.githubusercontent.com/1691518/219252132-cf08165e-77e5-4439-81fa-85c4370ae6e9.png">

**Describe the solution you'd like**

To support entering this `exclude` field. A sample setting value is:
```
{"vlan":5,"clusterNetwork":"abc","range":"192.168.77.0/24", "exclude": ["192.168.77.1/32",
  "192.168.77.254/32"]}
```

The type is a list of strings. Each element should be a valid CIDR.

**Describe alternatives you've considered**

**Additional context**

**Comment by Martin-Weiss** [comment]
Why do we need such an exclude list?
Can’t we use static IP addresses for the nodes in the storage network?
Or even better - can’t we use a completely not-routed network?

**Comment by bk201** [comment]
> Can’t we use static IP addresses for the nodes in the storage network? Or even better - can’t we use a completely not-routed network?

The IPs are allocated automatically with whereabouts cni. And yes depending on the configuration, it can reside in an isolate non-routed network.

**Comment by Martin-Weiss** [comment]
Thanks - then they question is “why do we need an exclude list”? Can’t we require an isolated network / VLAN?

**Comment by TachunLin** [comment]
Verifying **fixed** on `v1.1.2-rc1`. Close this issue. 

### Result
* We can set exclude list on storage-network settings
![image](https://user-images.githubusercontent.com/29251855/220813176-55c485a6-e4ff-4e08-b010-e865a860860e.png)

* Can successfully apply setting 
![image](https://user-images.githubusercontent.com/29251855/220829402-c724acc6-5fb8-4b23-9478-c688104900fa.png)

* Check the Harvester Storage Network setting's status is `True` and the type is `configured`
  ```
  harvester-t4mvc:~ # kubectl get settings.harvesterhci.io storage-network -o yaml
  apiVersion: harvesterhci.io/v1beta1
  kind: Setting
  metadata:
    annotations:
      storage-network.settings.harvesterhci.io/hash: ec5f8b96b0ac32b461b21315eb8b27fab31af9b2
      storage-network.settings.harvesterhci.io/net-attach-def: harvester-system/storagenetwork-fxfj7
      storage-network.settings.harvesterhci.io/old-net-attach-def: ""
    creationTimestamp: "2023-02-20T04:29:39Z"
    generation: 13
    name: storage-network
    resourceVersion: "4204343"
    uid: 26f3e42e-cb91-4f47-a1a9-85cccf6738ef
  status:
    conditions:
    - lastUpdateTime: "2023-02-23T03:09:49Z"
      reason: Completed
      status: "True"
      type: configured
  value: '{"vlan":1216,"clusterNetwork":"cn","range":"10.84.99.0/24","exclude":["10.84.99.10/32"]}'
  ```

* Check all Longhorn instance-manager-e and instance-manager-r are ready and networks are correct.
* Check the Annotations k8s.v1.cni.cncf.io/network-status has an interface named lhnet1 and the IP address is in the IP range.

  ```
  harvester-t4mvc:~ # kubectl get pods -n longhorn-system -l longhorn.io/component=instance-manager -o yaml
  apiVersion: v1
  items:
  - apiVersion: v1
    kind: Pod
    metadata:
      annotations:
        cni.projectcalico.org/containerID: e1ceac11a0dfaaae61ce41ac495de7459be695d662caf7b51c0804cc367893dc
        cni.projectcalico.org/podIP: 10.52.2.19/32
        cni.projectcalico.org/podIPs: 10.52.2.19/32
        k8s.v1.cni.cncf.io/network-status: |-
          [{
              "name": "k8s-pod-network",
              "ips": [
                  "10.52.2.19"
              ],
              "default": true,
              "dns": {}
          },{
              "name": "harvester-system/storagenetwork-fxfj7",
              "interface": "lhnet1",
              "ips": [
                  "10.84.99.5"
              ],
              "mac": "9e:75:a4:2f:50:fb",
              "dns": {}
          }]
        k8s.v1.cni.cncf.io/networks: '[{"namespace": "harvester-system", "name": "storagenetwork-fxfj7",
          "interface": "lhnet1"}]'
        k8s.v1.cni.cncf.io/networks-status: |-
          [{
              "name": "k8s-pod-network",
              "ips": [
                  "10.52.2.19"
              ],
              "default": true,
              "dns": {}
          },{
              "name": "harvester-system/storagenetwork-fxfj7",
              "interface": "lhnet1",
              "ips": [
                  "10.84.99.5"
              ],
              "mac": "9e:75:a4:2f:50:fb",
              "dns": {}
          }]
        kubernetes.io/psp: global-unrestricted-psp
        longhorn.io/last-applied-tolerations: '[{"key":"kubevirt.io/drain","operator":"Exists","effect":"NoSchedule"}]'
  ```

* Check can ping the ip assigned to storage-network 
  ```
  harvester-t4mvc:~ # kubectl exec -it instance-manager-e-15f0ad51 -n longhorn-system -- /bin/bash
  instance-manager-e-15f0ad51:/ # ping 10.84.99.5
  PING 10.84.99.5 (10.84.99.5) 56(84) bytes of data.
  64 bytes from 10.84.99.5: icmp_seq=1 ttl=64 time=0.036 ms
  64 bytes from 10.84.99.5: icmp_seq=2 ttl=64 time=0.040 ms
  64 bytes from 10.84.99.5: icmp_seq=3 ttl=64 time=0.044 ms
  64 bytes from 10.84.99.5: icmp_seq=4 ttl=64 time=0.040 ms
  
  ```

* Check can't ping the ip set in the excluded list of storage-network 
  ![image](https://user-images.githubusercontent.com/29251855/220835232-f1628d3c-dbc5-49a4-ae9d-043cae8bc4d3.png)

  ```
  harvester-t4mvc:~ # kubectl exec -it instance-manager-e-15f0ad51 -n longhorn-system -- /bin/bash
  instance-manager-e-15f0ad51:/ # ping 10.84.99.4
  PING 10.84.99.4 (10.84.99.4) 56(84) bytes of data.
  From 10.84.99.3 icmp_seq=1 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=5 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=6 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=7 Destination Host Unreachable
  ^C
  --- 10.84.99.4 ping statistics ---
  8 packets transmitted, 0 received, +4 errors, 100% packet loss, time 7122ms
  pipe 4
  instance-manager-e-15f0ad51:/ # ping 10.84.99.5
  PING 10.84.99.5 (10.84.99.5) 56(84) bytes of data.
  From 10.84.99.3 icmp_seq=1 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=2 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=3 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=4 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=5 Destination Host Unreachable
  From 10.84.99.3 icmp_seq=6 Destination Host Unreachable
  ^C
  --- 10.84.99.5 ping statistics ---
  7 packets transmitted, 0 received, +6 errors, 100% packet loss, time 6146ms
  pipe 3
  instance-manager-e-15f0ad51:/ # ping 10.84.99.3
  PING 10.84.99.3 (10.84.99.3) 56(84) bytes of data.
  64 bytes from 10.84.99.3: icmp_seq=1 ttl=64 time=0.034 ms
  64 bytes from 10.84.99.3: icmp_seq=2 ttl=64 time=0.039 ms
  64 bytes from 10.84.99.3: icmp_seq=3 ttl=64 time=0.044 ms
  ^C
  --- 10.84.99.3 ping statistics ---
  3 packets transmitted, 3 received, 0% packet loss, time 2034ms
  rtt min/avg/max/mdev = 0.034/0.039/0.044/0.004 ms
  ```

### Test Information
* Test Environment: 4 nodes Harvester bare machines
* Harvester version: v1.1.2-rc1

### Verify Steps
1. Set `ui-source` settings to `external` and refresh page 
2. Create a cluster network `cn`
3. Create a vlan config `nc` and set `ens1f1` for the uplink NIC
4. Create an `Active` vlan netowork with id `1216`
5. Enable storage-network in settings
6. Set vlan id: `1216`, cluster network `cn` , IP range and exclude list 
![image](https://user-images.githubusercontent.com/29251855/220836571-74527da1-9f3a-437e-a59a-7283eca935df.png)

7. ssh to management node and change to root 
8. Execute `kubectl get settings.harvesterhci.io storage-network -o yaml` 
9. Execute `kubectl get pods -n longhorn-system -l longhorn.io/component=instance-manager -o yaml` 
10. Get all `instance-manager-e` or `instance-manager-r` pod list
  ```
  kubectl get pods -A | grep instance-manager-e
  ```
11. Access the  one of the instance-manager-e pod 
  ```
  kubectl exec -it instance-manager-e-15f0ad51 -n longhorn-system -- /bin/bash
  ```
12. ping the ip address assigned to storage network 
13. ping the ip address excluded on the storage network

**Comment by bk201** [comment]
Reopen because this change causes a regression if we don't provide any IP in the "Exclude" field, for example:
<img width="457" alt="Screenshot 2023-02-24 at 4 26 25 PM" src="https://user-images.githubusercontent.com/1691518/221129960-4cad9e90-6dab-44ab-a788-d2ab7d8d67eb.png">

This produces the setting which `exclude` is a list of empty string:

```
'{"vlan":5,"clusterNetwork":"aa","range":"192.168.66.0/24","exclude":[""]}'
```

And instance-manager-e/r pod will fail to start because of:
```
Events:
  Type     Reason                  Age                  From     Message
  ----     ------                  ----                 ----     -------
  Normal   AddedInterface          3m37s                multus   Add eth0 [10.52.0.91/32] from k8s-pod-network
  Warning  FailedCreatePodSandBox  3m37s                kubelet  Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "f21bbc16ee5f81fe42b99f62
2c016bbbf560230d8b9f8b46438002a24b45c126": plugin type="multus" name="multus-cni-network" failed (add): [longhorn-system/instance-manager-e-e72d2ab5/541f2edf-0681-4e15-9658-db53c95807e2:sto
ragenetwork-gzjwg]: error adding container to network "storagenetwork-gzjwg": invalid CIDR in exclude list : invalid CIDR address:
  Normal   SandboxChanged          3s (x18 over 3m37s)  kubelet  Pod sandbox changed, it will be killed and re-created.
```

We need the `exclude` to be an empty list or not presented at all if the user doesn't specify anything. (Like `exclude: []`). @WuJun2016 please help take a look.

**Comment by alexdepalex** [comment]
> Thanks - then they question is “why do we need an exclude list”? Can’t we require an isolated network / VLAN?

Not every environment allows non-routable networks. This mostly results in networks with a gateway and pre-allocated IP's for maintenance, that need to be excluded from the storage network dhcp range.

**Comment by TachunLin** [comment]
#### Verify when we set empty exclude list on storage-network can work as expected

* When we don't provide exclude list in the storage network
![image](https://user-images.githubusercontent.com/29251855/222424615-2bce8f51-09f7-4da7-bacd-752e98e0b149.png)

* The produced setting will hide and not present the exclude list  
```
{
  "vlan": 1216,
  "clusterNetwork": "cn",
  "range": "10.84.99.0/24"
}

harvester-qqvjm:~ # kubectl get settings.harvesterhci.io storage-network -o yaml
apiVersion: harvesterhci.io/v1beta1
kind: Setting
metadata:
  annotations:
    storage-network.settings.harvesterhci.io/hash: cd3e42878a5a73d73ecffe8127c75230114fe825
    storage-network.settings.harvesterhci.io/net-attach-def: harvester-system/storagenetwork-56247
    storage-network.settings.harvesterhci.io/old-net-attach-def: ""
  creationTimestamp: "2023-03-02T11:27:42Z"
  generation: 10
  name: storage-network
  resourceVersion: "54471"
  uid: 30d58764-2168-49a6-a333-bede7f13eb51
status:
  conditions:
  - lastUpdateTime: "2023-03-02T12:08:43Z"
    reason: Completed
    status: "True"
    type: configured
value: '{"vlan":1216,"clusterNetwork":"cn","range":"10.84.99.0/24"}'
```

* All instance-manager-e/r pod keep in Running state
```
harvester-qqvjm:~ # kubectl get pods -A | grep instance-manager
longhorn-system             instance-manager-e-19d77410                               1/1     Running     0             33m
longhorn-system             instance-manager-e-211d5615                               1/1     Running     0             33m
longhorn-system             instance-manager-e-8b1c7297                               1/1     Running     0             33m
longhorn-system             instance-manager-e-f96dda0a                               1/1     Running     0             33m
longhorn-system             instance-manager-r-751ce3e3                               1/1     Running     0             33m
longhorn-system             instance-manager-r-9aee339f                               1/1     Running     0             32m
longhorn-system             instance-manager-r-eab9aadc                               1/1     Running     0             33m
longhorn-system             instance-manager-r-f8645a98                               1/1     Running     0             33
```





* Check the Annotations k8s.v1.cni.cncf.io/network-status has an interface named lhnet1 and the IP address is in the IP range.
```
k8s.v1.cni.cncf.io/networks: '[{"namespace": "harvester-system", "name": "storagenetwork-56247",
        "interface": "lhnet1"}]'
      k8s.v1.cni.cncf.io/networks-status: |-
        [{
            "name": "k8s-pod-network",
            "ips": [
                "10.52.1.18"
            ],
            "default": true,
            "dns": {}
        },{
            "name": "harvester-system/storagenetwork-56247",
            "interface": "lhnet1",
            "ips": [
                "10.84.99.2"
            ],
            "mac": "a6:82:50:1d:79:40",
            "dns": {}
        }]
```

**Comment by TachunLin** [comment]
#### Verify the exclude list set on storage-network can work as expected

* The produced setting written in the exclude list 
```
value: '{"vlan":1216,"clusterNetwork":"cn","range":"10.84.99.0/24","exclude":["10.84.99.5/32","10.84.99.4/32"]}'
```
* All instance-manager-e/r pod keep in Running state
```
harvester-qqvjm:~ # kubectl get pods -A | grep instance-manager
longhorn-system             instance-manager-e-19d77410                               1/1     Running     0              2m51s
longhorn-system             instance-manager-e-211d5615                               1/1     Running     0              2m52s
longhorn-system             instance-manager-e-8b1c7297                               1/1     Running     0              2m52s
longhorn-system             instance-manager-e-f96dda0a                               1/1     Running     0              2m51s
longhorn-system             instance-manager-r-751ce3e3                               1/1     Running     0              2m51s
longhorn-system             instance-manager-r-9aee339f                               1/1     Running     0              2m51s
longhorn-system             instance-manager-r-eab9aadc                               1/1     Running     0              2m52s
longhorn-system             instance-manager-r-f8645a98                               1/1     Running     0              2m51s
```

* Check can't ping the ip set in the excluded list of storage-network 
```
harvester-qqvjm:~ # kubectl exec -it instance-manager-e-19d77410 -n longhorn-system -- /bin/bash
instance-manager-e-19d77410:/ # ping 10.84.99.4
PING 10.84.99.4 (10.84.99.4) 56(84) bytes of data.
From 10.84.99.8 icmp_seq=1 Destination Host Unreachable
From 10.84.99.8 icmp_seq=5 Destination Host Unreachable
From 10.84.99.8 icmp_seq=8 Destination Host Unreachable
From 10.84.99.8 icmp_seq=9 Destination Host Unreachable
^C
--- 10.84.99.4 ping statistics ---
10 packets transmitted, 0 received, +4 errors, 100% packet loss, time 9199ms
pipe 4
instance-manager-e-19d77410:/ # ping 10.84.99.5
PING 10.84.99.5 (10.84.99.5) 56(84) bytes of data.
From 10.84.99.8 icmp_seq=1 Destination Host Unreachable
From 10.84.99.8 icmp_seq=5 Destination Host Unreachable
From 10.84.99.8 icmp_seq=6 Destination Host Unreachable

```

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.1.2-rc1` with latest v1.1 ui-index . Close this issue.

### Result

- [x] Check the empty exclude list in storage network can work as expected 
https://github.com/harvester/harvester/issues/3478#issuecomment-1451811431
- [x] Check when we set exclude list range in storage network can work as expected 
https://github.com/harvester/harvester/issues/3478#issuecomment-1451850317

### Test Information
* Test Environment: 4 nodes Harvester bare machines
* Harvester version: `v1.1.2-rc1`
* ui-index: https://releases.rancher.com/harvester-ui/dashboard/release-harvester-v1.1/index.html
* ui-source: external

### Verify Steps
https://github.com/harvester/harvester/issues/3478#issuecomment-1441246424
